File: lib/ace/playbook.ts
import { Bullet, Playbook, PlaybookStats, DeltaOperation, DeltaBatch } from './types'

export class PlaybookImpl implements Playbook {
  bullets: Bullet[] = []
  sections: string[] = ['defaults', 'guidelines', 'examples', 'warnings']

  constructor(initialBullets: Bullet[] = []) {
    this.bullets = initialBullets
  }

  stats(): PlaybookStats {
    const total_bullets = this.bullets.length
    const helpful_bullets = this.bullets.filter(b => b.metadata.helpful > b.metadata.harmful).length
    const harmful_bullets = this.bullets.filter(b => b.metadata.harmful > b.metadata.helpful).length
    
    const sections: { [section: string]: number } = {}
    this.sections.forEach(section => {
      sections[section] = this.bullets.filter(b => b.section === section).length
    })

    return {
      total_bullets,
      helpful_bullets,
      harmful_bullets,
      sections
    }
  }

  as_prompt(): string {
    if (this.bullets.length === 0) {
      return ''
    }

    const sections: { [key: string]: Bullet[] } = {}
    this.sections.forEach(section => {
      sections[section] = this.bullets.filter(b => b.section === section)
    })

    let prompt = ''
    this.sections.forEach(section => {
      const bullets = sections[section]
      if (bullets.length > 0) {
        prompt += `\n## ${section.toUpperCase()}\n`
        bullets.forEach(bullet => {
          const helpful = bullet.metadata.helpful || 0
          const harmful = bullet.metadata.harmful || 0
          const score = helpful - harmful
          const scoreStr = score > 0 ? `+${score}` : score.toString()
          prompt += `[${bullet.id}] ${bullet.content} (${scoreStr})\n`
        })
      }
    })

    return prompt.trim()
  }

  get_bullet(id: string): Bullet | null {
    return this.bullets.find(b => b.id === id) || null
  }

  tag_bullet(id: string, tag: string): void {
    const bullet = this.get_bullet(id)
    if (bullet && !bullet.tags.includes(tag)) {
      bullet.tags.push(tag)
    }
  }

  apply_delta(delta: DeltaBatch): void {
    delta.operations.forEach(op => {
      switch (op.type) {
        case 'ADD':
          const newBullet: Bullet = {
            id: op.bullet_id || `bullet_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
            content: op.content,
            section: op.section,
            metadata: {
              helpful: 0,
              harmful: 0,
              ...op.metadata
            },
            tags: []
          }
          this.bullets.push(newBullet)
          break

        case 'UPDATE':
          const bullet = this.get_bullet(op.bullet_id || '')
          if (bullet) {
            bullet.content = op.content
            bullet.section = op.section
            bullet.metadata = { ...bullet.metadata, ...op.metadata }
          }
          break

        case 'DELETE':
          this.bullets = this.bullets.filter(b => b.id !== op.bullet_id)
          break
      }
    })
  }
}

export class DeltaBatchImpl implements DeltaBatch {
  operations: DeltaOperation[] = []

  constructor(operations: DeltaOperation[] = []) {
    this.operations = operations
  }

  static from_json(data: any): DeltaBatch {
    const operations: DeltaOperation[] = []
    
    if (data.operations && Array.isArray(data.operations)) {
      data.operations.forEach((op: any) => {
        if (op.type && op.section && op.content) {
          operations.push({
            type: op.type,
            section: op.section,
            content: op.content,
            metadata: op.metadata || {},
            bullet_id: op.bullet_id
          })
        }
      })
    }

    return new DeltaBatchImpl(operations)
  }
}

File: ACE-open/ace/playbook.py
"""Playbook storage and mutation logic for ACE."""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from typing import Dict, Iterable, List, Optional

from .delta import DeltaBatch, DeltaOperation


@dataclass
class Bullet:
    """Single playbook entry."""

    id: str
    section: str
    content: str
    helpful: int = 0
    harmful: int = 0
    neutral: int = 0
    created_at: str = field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )
    updated_at: str = field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )

    def apply_metadata(self, metadata: Dict[str, int]) -> None:
        for key, value in metadata.items():
            if hasattr(self, key):
                setattr(self, key, int(value))

    def tag(self, tag: str, increment: int = 1) -> None:
        if tag not in ("helpful", "harmful", "neutral"):
            raise ValueError(f"Unsupported tag: {tag}")
        current = getattr(self, tag)
        setattr(self, tag, current + increment)
        self.updated_at = datetime.now(timezone.utc).isoformat()


class Playbook:
    """Structured context store as defined by ACE."""

    def __init__(self) -> None:
        self._bullets: Dict[str, Bullet] = {}
        self._sections: Dict[str, List[str]] = {}
        self._next_id = 0

    # ------------------------------------------------------------------ #
    # CRUD utils
    # ------------------------------------------------------------------ #
    def add_bullet(
        self,
        section: str,
        content: str,
        bullet_id: Optional[str] = None,
        metadata: Optional[Dict[str, int]] = None,
    ) -> Bullet:
        bullet_id = bullet_id or self._generate_id(section)
        metadata = metadata or {}
        bullet = Bullet(id=bullet_id, section=section, content=content)
        bullet.apply_metadata(metadata)
        self._bullets[bullet_id] = bullet
        self._sections.setdefault(section, []).append(bullet_id)
        return bullet

    def update_bullet(
        self,
        bullet_id: str,
        *,
        content: Optional[str] = None,
        metadata: Optional[Dict[str, int]] = None,
    ) -> Optional[Bullet]:
        bullet = self._bullets.get(bullet_id)
        if bullet is None:
            return None
        if content is not None:
            bullet.content = content
        if metadata:
            bullet.apply_metadata(metadata)
        bullet.updated_at = datetime.now(timezone.utc).isoformat()
        return bullet

    def tag_bullet(self, bullet_id: str, tag: str, increment: int = 1) -> Optional[Bullet]:
        bullet = self._bullets.get(bullet_id)
        if bullet is None:
            return None
        bullet.tag(tag, increment=increment)
        return bullet

    def remove_bullet(self, bullet_id: str) -> None:
        bullet = self._bullets.pop(bullet_id, None)
        if bullet is None:
            return
        section_list = self._sections.get(bullet.section)
        if section_list:
            self._sections[bullet.section] = [
                bid for bid in section_list if bid != bullet_id
            ]
            if not self._sections[bullet.section]:
                del self._sections[bullet.section]

    def get_bullet(self, bullet_id: str) -> Optional[Bullet]:
        return self._bullets.get(bullet_id)

    def bullets(self) -> List[Bullet]:
        return list(self._bullets.values())

    # ------------------------------------------------------------------ #
    # Serialization
    # ------------------------------------------------------------------ #
    def to_dict(self) -> Dict[str, object]:
        return {
            "bullets": {bullet_id: asdict(bullet) for bullet_id, bullet in self._bullets.items()},
            "sections": self._sections,
            "next_id": self._next_id,
        }

    @classmethod
    def from_dict(cls, payload: Dict[str, object]) -> "Playbook":
        instance = cls()
        bullets_payload = payload.get("bullets", {})
        if isinstance(bullets_payload, dict):
            for bullet_id, bullet_value in bullets_payload.items():
                if isinstance(bullet_value, dict):
                    instance._bullets[bullet_id] = Bullet(**bullet_value)
        sections_payload = payload.get("sections", {})
        if isinstance(sections_payload, dict):
            instance._sections = {
                section: list(ids) if isinstance(ids, Iterable) else []
                for section, ids in sections_payload.items()
            }
        instance._next_id = int(payload.get("next_id", 0))
        return instance

    def dumps(self) -> str:
        return json.dumps(self.to_dict(), ensure_ascii=False, indent=2)

    @classmethod
    def loads(cls, data: str) -> "Playbook":
        payload = json.loads(data)
        if not isinstance(payload, dict):
            raise ValueError("Playbook serialization must be a JSON object.")
        return cls.from_dict(payload)

    # ------------------------------------------------------------------ #
    # Delta application
    # ------------------------------------------------------------------ #
    def apply_delta(self, delta: DeltaBatch) -> None:
        for operation in delta.operations:
            self._apply_operation(operation)

    def _apply_operation(self, operation: DeltaOperation) -> None:
        op_type = operation.type.upper()
        if op_type == "ADD":
            self.add_bullet(
                section=operation.section,
                content=operation.content or "",
                bullet_id=operation.bullet_id,
                metadata=operation.metadata,
            )
        elif op_type == "UPDATE":
            if operation.bullet_id is None:
                return
            self.update_bullet(
                operation.bullet_id,
                content=operation.content,
                metadata=operation.metadata,
            )
        elif op_type == "TAG":
            if operation.bullet_id is None:
                return
            for tag, increment in operation.metadata.items():
                self.tag_bullet(operation.bullet_id, tag, increment)
        elif op_type == "REMOVE":
            if operation.bullet_id is None:
                return
            self.remove_bullet(operation.bullet_id)

    # ------------------------------------------------------------------ #
    # Presentation helpers
    # ------------------------------------------------------------------ #
    def as_prompt(self) -> str:
        """Return a human-readable playbook string for prompting LLMs."""
        parts: List[str] = []
        for section, bullet_ids in sorted(self._sections.items()):
            parts.append(f"## {section}")
            for bullet_id in bullet_ids:
                bullet = self._bullets[bullet_id]
                counters = (
                    f"(helpful={bullet.helpful}, harmful={bullet.harmful}, neutral={bullet.neutral})"
                )
                parts.append(f"- [{bullet.id}] {bullet.content} {counters}")
        return "\n".join(parts)

    def stats(self) -> Dict[str, object]:
        return {
            "sections": len(self._sections),
            "bullets": len(self._bullets),
            "tags": {
                "helpful": sum(b.helpful for b in self._bullets.values()),
                "harmful": sum(b.harmful for b in self._bullets.values()),
                "neutral": sum(b.neutral for b in self._bullets.values()),
            },
        }

    # ------------------------------------------------------------------ #
    # Internal helpers
    # ------------------------------------------------------------------ #
    def _generate_id(self, section: str) -> str:
        self._next_id += 1
        section_prefix = section.split()[0].lower()
        return f"{section_prefix}-{self._next_id:05d}"
